File: ts-sdk/src/t-ibe/bfibe-bls12381-shortpk-otp-hmac.ts
// Copyright (c) Aptos Labs
// SPDX-License-Identifier: Apache-2.0

// A threshold IBE based on Boneh-Franklin IBE, where...
// - The underlying curve is BLS12-381.
// - The public key is in G1.
// - The symmetric cipher inside is a one-time pad.
// - HMAC-SHA3-256 is used for authentication.
// - In decryption, decryption key shares can together reconstruct and then it will be like normal Boneh-Franklin IBE decryption.

import { Deserializer, Serializer } from "@aptos-labs/ts-sdk";
import { Fp2 } from "@noble/curves/abstract/tower";
import { WeierstrassPoint } from "@noble/curves/abstract/weierstrass";
import { bls12_381 } from "@noble/curves/bls12-381";
import { bytesToNumberBE, bytesToNumberLE, numberToBytesLE } from "@noble/curves/utils";
import { bytesToHex, hexToBytes, randomBytes } from "@noble/hashes/utils";
import { frMod, frMul, frInv } from "../group/bls12381g1";
import { hmac_sha3_256, kdf, xorBytes } from "../utils";
import { Result } from "../result";

const DST_OTP = new TextEncoder().encode("BONEH_FRANKLIN_BLS12381_SHORT_PK/OTP");
const DST_ID_HASH = new TextEncoder().encode("BONEH_FRANKLIN_BLS12381_SHORT_PK/HASH_ID_TO_CURVE");
const DST_MAC = new TextEncoder().encode("BONEH_FRANKLIN_BLS12381_SHORT_PK/MAC");

export class MasterPublicKey {
    basePoint: WeierstrassPoint<bigint>;
    pk: WeierstrassPoint<bigint>;
    constructor(basePoint: WeierstrassPoint<bigint>, pk: WeierstrassPoint<bigint>) {
        this.basePoint = basePoint;
        this.pk = pk;
    }

    static deserialize(deserializer: Deserializer): Result<MasterPublicKey> {
        const task = (_extra: Record<string, any>) => {
            const baseBytes = deserializer.deserializeBytes();
            const base = bls12_381.G1.Point.fromBytes(baseBytes) as unknown as WeierstrassPoint<bigint>;
            const pkBytes = deserializer.deserializeBytes();
            const pk = bls12_381.G1.Point.fromBytes(pkBytes) as unknown as WeierstrassPoint<bigint>;
            return new MasterPublicKey(base, pk);
        };
        return Result.capture({task, recordsExecutionTimeMs: false});
    }

    serialize(serializer: Serializer): void {
        serializer.serializeBytes((this.basePoint as any).toBytes());
        serializer.serializeBytes((this.pk as any).toBytes());
    }

    static fromBytes(bytes: Uint8Array): Result<MasterPublicKey> {
        const task = (_extra: Record<string, any>) => {
            const des = new Deserializer(bytes);
            const ret = MasterPublicKey.deserialize(des).unwrapOrThrow('MasterPublicKey.fromBytes failed');
            if (des.remaining() !== 0) throw 'MasterPublicKey.fromBytes: trailing bytes';
            return ret;
        };
        return Result.capture({task, recordsExecutionTimeMs: false});
    }

    static fromHex(hex: string): Result<MasterPublicKey> {
        const task = (_extra: Record<string, any>) =>
            MasterPublicKey.fromBytes(hexToBytes(hex)).unwrapOrThrow('MasterPublicKey.fromHex failed');
        return Result.capture({task, recordsExecutionTimeMs: false});
    }

    toBytes(): Uint8Array {
        const serializer = new Serializer();
        this.serialize(serializer);
        return serializer.toUint8Array();
    }

    toHex(): string {
        return bytesToHex(this.toBytes());
    }
}

export class MasterPrivateKey {
    base: WeierstrassPoint<bigint>;
    scalar: bigint;

    constructor(base: WeierstrassPoint<bigint>, scalar: bigint) {
        this.base = base;
        this.scalar = scalar;
    }

    static deserialize(deserializer: Deserializer): Result<MasterPrivateKey> {
        const task = (_extra: Record<string, any>) => {
            const baseBytes = deserializer.deserializeBytes();
            const base = bls12_381.G1.Point.fromBytes(baseBytes) as unknown as WeierstrassPoint<bigint>;
            const scalarBytes = deserializer.deserializeBytes();
            const scalar = bytesToNumberLE(scalarBytes);
            return new MasterPrivateKey(base, scalar);
        };
        return Result.capture({task, recordsExecutionTimeMs: false});
    }

    serialize(serializer: Serializer): void {
        serializer.serializeBytes((this.base as any).toBytes());
        serializer.serializeBytes(numberToBytesLE(this.scalar, 32));
    }

    static fromBytes(bytes: Uint8Array): Result<MasterPrivateKey> {
        const task = (_extra: Record<string, any>) => {
            const des = new Deserializer(bytes);
            const ret = MasterPrivateKey.deserialize(des).unwrapOrThrow('MasterPrivateKey.fromBytes failed');
            if (des.remaining() !== 0) throw 'MasterPrivateKey.fromBytes: trailing bytes';
            return ret;
        };
        return Result.capture({task, recordsExecutionTimeMs: false});
    }

    static fromHex(hex: string): Result<MasterPrivateKey> {
        const task = (_extra: Record<string, any>) =>
            MasterPrivateKey.fromBytes(hexToBytes(hex)).unwrapOrThrow('MasterPrivateKey.fromHex failed');
        return Result.capture({task, recordsExecutionTimeMs: false});
    }

    toBytes(): Uint8Array {
        const serializer = new Serializer();
        this.serialize(serializer);
        return serializer.toUint8Array();
    }

    toHex(): string {
        return bytesToHex(this.toBytes());
    }
}

export class IdentityDecryptionKeyShare {
    evalPoint: bigint;
    idkShare: WeierstrassPoint<Fp2>;
    proof: Uint8Array | undefined;

    constructor(evalPoint: bigint, idkShare: WeierstrassPoint<Fp2>, proof: Uint8Array | undefined) {
        this.evalPoint = evalPoint;
        this.idkShare = idkShare;
        this.proof = proof;
    }

    serialize(serializer: Serializer): void {
        serializer.serializeBytes(numberToBytesLE(this.evalPoint, 32));
        serializer.serializeBytes(this.idkShare.toBytes());
        serializer.serializeU8(this.proof !== undefined ? 1 : 0);
        if (this.proof !== undefined) {
            serializer.serializeBytes(this.proof);
        }
    }

    static deserialize(deserializer: Deserializer): Result<IdentityDecryptionKeyShare> {
        const task = (_extra: Record<string, any>) => {
            const evalPointBytes = deserializer.deserializeBytes();
            if (evalPointBytes.length !== 32) throw 'IdentityDecryptionKeyShare: expected 32-byte evalPoint';
            const evalPoint = bytesToNumberLE(evalPointBytes);
            const idkShareBytes = deserializer.deserializeBytes();
            const idkShare = bls12_381.G2.Point.fromBytes(idkShareBytes) as unknown as WeierstrassPoint<Fp2>;
            const hasProof = deserializer.deserializeU8() !== 0;
            const proof = hasProof ? deserializer.deserializeBytes() : undefined;
            return new IdentityDecryptionKeyShare(evalPoint, idkShare, proof);
        };
        return Result.capture({task, recordsExecutionTimeMs: false});
    }

    static fromBytes(bytes: Uint8Array): Result<IdentityDecryptionKeyShare> {
        const task = (_extra: Record<string, any>) => {
            const des = new Deserializer(bytes);
            const ret = IdentityDecryptionKeyShare.deserialize(des).unwrapOrThrow('IdentityDecryptionKeyShare.fromBytes failed');
            if (des.remaining() !== 0) throw 'IdentityDecryptionKeyShare.fromBytes: trailing bytes';
            return ret;
        };
        return Result.capture({task, recordsExecutionTimeMs: false});
    }

    static fromHex(hex: string): Result<IdentityDecryptionKeyShare> {
        const task = (_extra: Record<string, any>) =>
            IdentityDecryptionKeyShare.fromBytes(hexToBytes(hex)).unwrapOrThrow('IdentityDecryptionKeyShare.fromHex failed');
        return Result.capture({task, recordsExecutionTimeMs: false});
    }

    toBytes(): Uint8Array {
        const serializer = new Serializer();
        this.serialize(serializer);
        return serializer.toUint8Array();
    }

    toHex(): string {
        return bytesToHex(this.toBytes());
    }
}

export class Ciphertext {
    c0: WeierstrassPoint<bigint>;
    symmetricCiph: Uint8Array;
    mac: Uint8Array;

    constructor(c0: WeierstrassPoint<bigint>, symmetricCiph: Uint8Array, mac: Uint8Array) {
        this.c0 = c0;
        this.symmetricCiph = symmetricCiph;
        this.mac = mac;
    }

    static deserialize(deserializer: Deserializer): Result<Ciphertext> {
        const task = (_extra: Record<string, any>) => {
            const c0Bytes = deserializer.deserializeBytes();
            const c0 = bls12_381.G1.Point.fromBytes(c0Bytes) as unknown as WeierstrassPoint<bigint>;
            const symmetricCiph = deserializer.deserializeBytes();
            const mac = deserializer.deserializeBytes();
            return new Ciphertext(c0, symmetricCiph, mac);
        };
        return Result.capture({task, recordsExecutionTimeMs: false});
    }

    serialize(serializer: Serializer): void {
        serializer.serializeBytes((this.c0 as any).toBytes());
        serializer.serializeBytes(this.symmetricCiph);
        serializer.serializeBytes(this.mac);
    }

    static fromBytes(bytes: Uint8Array): Result<Ciphertext> {
        const task = (_extra: Record<string, any>) => {
            const des = new Deserializer(bytes);
            const ret = Ciphertext.deserialize(des).unwrapOrThrow('Ciphertext.fromBytes failed');
            if (des.remaining() !== 0) throw 'Ciphertext.fromBytes: trailing bytes';
            return ret;
        };
        return Result.capture({task, recordsExecutionTimeMs: false});
    }

    static fromHex(hex: string): Result<Ciphertext> {
        const task = (_extra: Record<string, any>) =>
            Ciphertext.fromBytes(hexToBytes(hex)).unwrapOrThrow('Ciphertext.fromHex failed');
        return Result.capture({task, recordsExecutionTimeMs: false});
    }

    toBytes(): Uint8Array {
        const serializer = new Serializer();
        this.serialize(serializer);
        return serializer.toUint8Array();
    }

    toHex(): string {
        return bytesToHex(this.toBytes());
    }
}

export function keygenForTesting(): MasterPrivateKey {
    const base = bls12_381.G1.hashToCurve(randomBytes(32)) as unknown as WeierstrassPoint<bigint>;
    const scalar = bytesToNumberBE(bls12_381.utils.randomSecretKey());
    return new MasterPrivateKey(base, scalar);
}

export function derivePublicKey(msk: MasterPrivateKey): MasterPublicKey {
    const pk = msk.base.multiply(msk.scalar);
    return new MasterPublicKey(msk.base, pk);
}

export function encryptWithRandomness(mpk: MasterPublicKey, id: Uint8Array, plaintext: Uint8Array, randomness: Uint8Array): Ciphertext {
    const r = bytesToNumberLE(randomness);
    const idPoint = bls12_381.G2.hashToCurve(id, { DST: DST_ID_HASH }) as unknown as WeierstrassPoint<Fp2>;
    const seedElement = bls12_381.pairing(mpk.pk.multiply(r), idPoint);
    const seed = bls12381GtReprNobleToAptos(bls12_381.fields.Fp12.toBytes(seedElement)).unwrapOrThrow('encryptWithRandomness: Gt conversion failed');
    const otp = kdf(seed, DST_OTP, plaintext.length);
    const macKey = kdf(seed, DST_MAC, 32);
    const symmetricCiph = xorBytes(otp, plaintext);
    const mac = hmac_sha3_256(macKey, symmetricCiph);
    const c0 = mpk.basePoint.multiply(r);
    return new Ciphertext(c0, symmetricCiph, mac);
}

export function encrypt({mpk, id, plaintext}: {mpk: MasterPublicKey, id: Uint8Array, plaintext: Uint8Array}): Result<Ciphertext> {
    const task = (_extra: Record<string, any>) => {
        const r = bytesToNumberBE(bls12_381.utils.randomSecretKey());
        return encryptWithRandomness(mpk, id, plaintext, numberToBytesLE(r, 32));
    };
    return Result.capture({task, recordsExecutionTimeMs: true});
}

export function decrypt({idkShares, ciphertext}: {idkShares: IdentityDecryptionKeyShare[], ciphertext: Ciphertext}): Result<Uint8Array> {
    const task = (_extra: Record<string, any>) => {
        if (idkShares.length === 0) throw 'decrypt: no IDK shares provided';

        // Step 1: Lagrange interpolation in the exponent (G2).
        // Each share is (x_i = evalPoint_i, H(id)^{s_i}).
        // Compute idkFull = Σ_i λ_i * H(id)^{s_i} where λ_i are Lagrange coefficients at x=0.
        const xs = idkShares.map(s => frMod(s.evalPoint));
        for (let i = 0; i < xs.length; i++) {
            for (let j = i + 1; j < xs.length; j++) {
                if (xs[i] === xs[j]) throw 'decrypt: duplicate evalPoint';
            }
        }

        const lambdas: bigint[] = xs.map((xi, i) => {
            let lambda = 1n;
            for (let j = 0; j < xs.length; j++) {
                if (i === j) continue;
                // λ_i = Π_{j≠i} (0 - x_j) / (x_i - x_j)  in Fr
                lambda = frMul(lambda, frMul(frMod(-xs[j]), frInv(frMod(xi - xs[j]))));
            }
            return lambda;
        });

        let idkFull: WeierstrassPoint<Fp2> | null = null;
        for (let i = 0; i < idkShares.length; i++) {
            if (lambdas[i] === 0n) continue;
            const scaled = idkShares[i].idkShare.multiply(lambdas[i]);
            idkFull = idkFull === null ? scaled : idkFull.add(scaled);
        }
        if (idkFull === null) throw 'decrypt: all Lagrange coefficients were zero';

        // Step 2: Standard BF-IBE decryption using the reconstructed identity key.
        // pair(c0, idkFull) = pair(base^r, H(id)^s) = pair(base, H(id))^{rs} = pair(pk^r, H(id))
        const seedElement = bls12_381.pairing(ciphertext.c0, idkFull);
        const seed = bls12381GtReprNobleToAptos(bls12_381.fields.Fp12.toBytes(seedElement)).unwrapOrThrow('decrypt: Gt conversion failed');
        const macKey = kdf(seed, DST_MAC, 32);
        const macAnother = hmac_sha3_256(macKey, ciphertext.symmetricCiph);
        if (bytesToHex(ciphertext.mac) !== bytesToHex(macAnother)) {
            throw 'decrypt: MAC verification failed';
        }
        const otp = kdf(seed, DST_OTP, ciphertext.symmetricCiph.length);
        return xorBytes(otp, ciphertext.symmetricCiph);
    };
    return Result.capture({task, recordsExecutionTimeMs: true});
}

/**
 * Aptos Gt format is defined in https://github.com/aptos-labs/aptos-core/blob/46d871fa1feb61ffafb73353a0755e8cc3aaed9d/aptos-move/framework/aptos-stdlib/sources/cryptography/bls12381_algebra.move#L204.
 * Noble format is similar except that each Fp element is big-endian.
 */
function bls12381GtReprNobleToAptos(noble: Uint8Array): Result<Uint8Array> {
    const task = (_extra: Record<string, any>) => {
        if (noble.length !== 576) {
            throw 'bls12381GtReprNobleToAptos failed with incorrect input length';
        }

        const chunks = [];
        for (let i = 0; i < noble.length; i += 48) {
            chunks.push(noble.slice(i, i + 48).reverse());
        }

        const result = new Uint8Array(576);
        for (let i = 0; i < 12; i++) {
            result.set(chunks[i], i * 48);
        }
        return result;
    };
    return Result.capture({task, recordsExecutionTimeMs: true});
}

File: python-sdk/src/ace_sdk/t_ibe/bfibe_bls12381_shortpk_otp_hmac.py
# Copyright (c) Aptos Labs
# SPDX-License-Identifier: Apache-2.0
"""
Mirrors src/t-ibe/bfibe-bls12381-shortpk-otp-hmac.ts.

Boneh-Franklin IBE over BLS12-381:
- The public key is in G1.
- Identity hashed to G2 (Q_id in G2, 96 bytes compressed).
- The symmetric cipher inside is a one-time pad.
- HMAC-SHA3-256 is used for authentication.
- In decryption, decryption key shares combine via Lagrange interpolation (G2).
"""

from __future__ import annotations

import hashlib

import py_ecc.optimized_bls12_381 as _ob
from py_ecc.bls.hash_to_curve import hash_to_G2

from ace_sdk.bcs import Deserializer, Serializer
from ace_sdk.group.bls12381_pairing import (
    fp12_eq,
    fp12_to_aptos_gt_bytes,
    g1_jacobian_to_affine,
    g2_jacobian_to_affine,
    pairing,
)
from ace_sdk.group.bls12381fr import fr_inv, fr_mod, fr_mul
from ace_sdk.group.bls12381g1 import _g1_from_compressed, _g1_to_compressed
from ace_sdk.group.bls12381g2 import _g2_from_compressed, _g2_to_compressed
from ace_sdk.result import Result
from ace_sdk.utils import hmac_sha3_256, kdf, rand_bytes, xor_bytes

DST_OTP = b"BONEH_FRANKLIN_BLS12381_SHORT_PK/OTP"
DST_ID_HASH = b"BONEH_FRANKLIN_BLS12381_SHORT_PK/HASH_ID_TO_CURVE"
DST_MAC = b"BONEH_FRANKLIN_BLS12381_SHORT_PK/MAC"

_G1Point = tuple
_G2Point = tuple


def _hash_id_to_g2(identity: bytes) -> _G2Point:
    return hash_to_G2(identity, DST_ID_HASH, hashlib.sha256)


def _seed_from_gt(g1_pt: _G1Point, g2_pt: _G2Point) -> bytes:
    e = pairing(g1_jacobian_to_affine(g1_pt), g2_jacobian_to_affine(g2_pt))
    return fp12_to_aptos_gt_bytes(e)


# == MasterPublicKey ==========================================================


class MasterPublicKey:
    def __init__(self, base_point: _G1Point, pk: _G1Point) -> None:
        self.base_point = base_point
        self.pk = pk

    @staticmethod
    def deserialize(deserializer: Deserializer) -> Result["MasterPublicKey"]:
        def task(_extra: dict) -> "MasterPublicKey":
            base_bytes = deserializer.deserialize_bytes()
            base = _g1_from_compressed(base_bytes)
            pk_bytes = deserializer.deserialize_bytes()
            pk = _g1_from_compressed(pk_bytes)
            return MasterPublicKey(base, pk)

        return Result.capture(task)

    def serialize(self, serializer: Serializer) -> None:
        serializer.serialize_bytes(_g1_to_compressed(self.base_point))
        serializer.serialize_bytes(_g1_to_compressed(self.pk))

    @staticmethod
    def from_bytes(data: bytes) -> Result["MasterPublicKey"]:
        def task(_extra: dict) -> "MasterPublicKey":
            deserializer = Deserializer(data)
            obj = MasterPublicKey.deserialize(deserializer).unwrap_or_throw(
                ValueError("MasterPublicKey.from_bytes failed")
            )
            if deserializer.remaining() != 0:
                raise ValueError("MasterPublicKey.from_bytes: trailing bytes")
            return obj

        return Result.capture(task)

    @staticmethod
    def from_hex(hex_str: str) -> Result["MasterPublicKey"]:
        def task(_extra: dict) -> "MasterPublicKey":
            return MasterPublicKey.from_bytes(bytes.fromhex(hex_str)).unwrap_or_throw(
                ValueError("MasterPublicKey.from_hex failed")
            )

        return Result.capture(task)

    def to_bytes(self) -> bytes:
        serializer = Serializer()
        self.serialize(serializer)
        return serializer.to_bytes()

    def to_hex(self) -> str:
        return self.to_bytes().hex()


# == MasterPrivateKey =========================================================


class MasterPrivateKey:
    def __init__(self, base: _G1Point, scalar: int) -> None:
        self.base = base
        self.scalar = scalar

    @staticmethod
    def deserialize(deserializer: Deserializer) -> Result["MasterPrivateKey"]:
        def task(_extra: dict) -> "MasterPrivateKey":
            base_bytes = deserializer.deserialize_bytes()
            base = _g1_from_compressed(base_bytes)
            scalar_bytes = deserializer.deserialize_bytes()
            scalar = int.from_bytes(scalar_bytes, "little")
            return MasterPrivateKey(base, scalar)

        return Result.capture(task)

    def serialize(self, serializer: Serializer) -> None:
        serializer.serialize_bytes(_g1_to_compressed(self.base))
        serializer.serialize_bytes(self.scalar.to_bytes(32, "little"))

    @staticmethod
    def from_bytes(data: bytes) -> Result["MasterPrivateKey"]:
        def task(_extra: dict) -> "MasterPrivateKey":
            deserializer = Deserializer(data)
            obj = MasterPrivateKey.deserialize(deserializer).unwrap_or_throw(
                ValueError("MasterPrivateKey.from_bytes failed")
            )
            if deserializer.remaining() != 0:
                raise ValueError("MasterPrivateKey.from_bytes: trailing bytes")
            return obj

        return Result.capture(task)

    @staticmethod
    def from_hex(hex_str: str) -> Result["MasterPrivateKey"]:
        def task(_extra: dict) -> "MasterPrivateKey":
            return MasterPrivateKey.from_bytes(bytes.fromhex(hex_str)).unwrap_or_throw(
                ValueError("MasterPrivateKey.from_hex failed")
            )

        return Result.capture(task)

    def to_bytes(self) -> bytes:
        serializer = Serializer()
        self.serialize(serializer)
        return serializer.to_bytes()

    def to_hex(self) -> str:
        return self.to_bytes().hex()


# == IdentityDecryptionKeyShare ================================================


class IdentityDecryptionKeyShare:
    def __init__(self, eval_point: int, idk_share: _G2Point, proof: bytes | None) -> None:
        self.eval_point = eval_point
        self.idk_share = idk_share
        self.proof = proof

    def serialize(self, serializer: Serializer) -> None:
        serializer.serialize_bytes(self.eval_point.to_bytes(32, "little"))
        serializer.serialize_bytes(_g2_to_compressed(self.idk_share))
        serializer.serialize_u8(1 if self.proof is not None else 0)
        if self.proof is not None:
            serializer.serialize_bytes(self.proof)

    @staticmethod
    def deserialize(deserializer: Deserializer) -> Result["IdentityDecryptionKeyShare"]:
        def task(_extra: dict) -> "IdentityDecryptionKeyShare":
            eval_point_bytes = deserializer.deserialize_bytes()
            if len(eval_point_bytes) != 32:
                raise ValueError("IdentityDecryptionKeyShare: expected 32-byte evalPoint")
            eval_point = int.from_bytes(eval_point_bytes, "little")
            idk_share_bytes = deserializer.deserialize_bytes()
            idk_share = _g2_from_compressed(idk_share_bytes)
            has_proof = deserializer.deserialize_u8() != 0
            proof = deserializer.deserialize_bytes() if has_proof else None
            return IdentityDecryptionKeyShare(eval_point, idk_share, proof)

        return Result.capture(task)

    @staticmethod
    def from_bytes(data: bytes) -> Result["IdentityDecryptionKeyShare"]:
        def task(_extra: dict) -> "IdentityDecryptionKeyShare":
            deserializer = Deserializer(data)
            obj = IdentityDecryptionKeyShare.deserialize(deserializer).unwrap_or_throw(
                ValueError("IdentityDecryptionKeyShare.from_bytes failed")
            )
            if deserializer.remaining() != 0:
                raise ValueError("IdentityDecryptionKeyShare.from_bytes: trailing bytes")
            return obj

        return Result.capture(task)

    @staticmethod
    def from_hex(hex_str: str) -> Result["IdentityDecryptionKeyShare"]:
        def task(_extra: dict) -> "IdentityDecryptionKeyShare":
            return IdentityDecryptionKeyShare.from_bytes(
                bytes.fromhex(hex_str)
            ).unwrap_or_throw(ValueError("IdentityDecryptionKeyShare.from_hex failed"))

        return Result.capture(task)

    def to_bytes(self) -> bytes:
        serializer = Serializer()
        self.serialize(serializer)
        return serializer.to_bytes()

    def to_hex(self) -> str:
        return self.to_bytes().hex()


# == Ciphertext ================================================================


class Ciphertext:
    def __init__(self, c0: _G1Point, symmetric_ciph: bytes, mac: bytes) -> None:
        self.c0 = c0
        self.symmetric_ciph = symmetric_ciph
        self.mac = mac

    @staticmethod
    def deserialize(deserializer: Deserializer) -> Result["Ciphertext"]:
        def task(_extra: dict) -> "Ciphertext":
            c0_bytes = deserializer.deserialize_bytes()
            c0 = _g1_from_compressed(c0_bytes)
            symmetric_ciph = deserializer.deserialize_bytes()
            mac = deserializer.deserialize_bytes()
            return Ciphertext(c0, symmetric_ciph, mac)

        return Result.capture(task)

    def serialize(self, serializer: Serializer) -> None:
        serializer.serialize_bytes(_g1_to_compressed(self.c0))
        serializer.serialize_bytes(self.symmetric_ciph)
        serializer.serialize_bytes(self.mac)

    @staticmethod
    def from_bytes(data: bytes) -> Result["Ciphertext"]:
        def task(_extra: dict) -> "Ciphertext":
            deserializer = Deserializer(data)
            obj = Ciphertext.deserialize(deserializer).unwrap_or_throw(
                ValueError("Ciphertext.from_bytes failed")
            )
            if deserializer.remaining() != 0:
                raise ValueError("Ciphertext.from_bytes: trailing bytes")
            return obj

        return Result.capture(task)

    @staticmethod
    def from_hex(hex_str: str) -> Result["Ciphertext"]:
        def task(_extra: dict) -> "Ciphertext":
            return Ciphertext.from_bytes(bytes.fromhex(hex_str)).unwrap_or_throw(
                ValueError("Ciphertext.from_hex failed")
            )

        return Result.capture(task)

    def to_bytes(self) -> bytes:
        serializer = Serializer()
        self.serialize(serializer)
        return serializer.to_bytes()

    def to_hex(self) -> str:
        return self.to_bytes().hex()


# == Key generation + derivation ==============================================


# Default G1 hash-to-curve DST used by @noble/curves' bls12_381.G1.htfDefaults
# (ts-sdk calls `bls12_381.G1.hashToCurve(randomBytes(32))` with no explicit DST).
_DEFAULT_G1_HTF_DST = b"BLS_SIG_BLS12381G1_XMD:SHA-256_SSWU_RO_NUL_"


def keygen_for_testing() -> MasterPrivateKey:
    """Generate a master private key with a hash-derived G1 base point. Tests only —
    in production the base point comes from the on-chain DKG session."""
    from py_ecc.bls.hash_to_curve import hash_to_G1

    base = hash_to_G1(rand_bytes(32), _DEFAULT_G1_HTF_DST, hashlib.sha256)
    scalar = fr_mod(int.from_bytes(rand_bytes(32), "big"))
    return MasterPrivateKey(base, scalar)


def derive_public_key(msk: MasterPrivateKey) -> MasterPublicKey:
    pk = _ob.multiply(msk.base, msk.scalar)
    return MasterPublicKey(msk.base, pk)


# == Encrypt / Decrypt =========================================================


def encrypt_with_randomness(
    mpk: MasterPublicKey, identity: bytes, plaintext: bytes, randomness: bytes
) -> Ciphertext:
    r = int.from_bytes(randomness, "little")
    id_point = _hash_id_to_g2(identity)
    seed_element_pt = _ob.multiply(mpk.pk, r)
    seed = _seed_from_gt(seed_element_pt, id_point)
    otp = kdf(seed, DST_OTP, len(plaintext))
    mac_key = kdf(seed, DST_MAC, 32)
    symmetric_ciph = xor_bytes(otp, plaintext)
    mac = hmac_sha3_256(mac_key, symmetric_ciph)
    c0 = _ob.multiply(mpk.base_point, r)
    return Ciphertext(c0, symmetric_ciph, mac)


def encrypt(mpk: MasterPublicKey, identity: bytes, plaintext: bytes) -> Result[Ciphertext]:
    def task(_extra: dict) -> Ciphertext:
        r = fr_mod(int.from_bytes(rand_bytes(64), "big"))
        randomness = r.to_bytes(32, "little")
        return encrypt_with_randomness(mpk, identity, plaintext, randomness)

    return Result.capture(task)


def verify_share(
    base_point: _G1Point, share_pk: _G1Point, identity: bytes, share: IdentityDecryptionKeyShare
) -> bool:
    """Verify that `share.idk_share = H_G2(id)^{f(x)}` where `share_pk = base_point^{f(x)}`.

    Pairing check: `e(base_point, idk_share) == e(share_pk, H_G2(id))`.
    """
    id_point = _hash_id_to_g2(identity)
    lhs = pairing(g1_jacobian_to_affine(base_point), g2_jacobian_to_affine(share.idk_share))
    rhs = pairing(g1_jacobian_to_affine(share_pk), g2_jacobian_to_affine(id_point))
    return fp12_eq(lhs, rhs)


def decrypt(
    idk_shares: list[IdentityDecryptionKeyShare], ciphertext: Ciphertext
) -> Result[bytes]:
    def task(_extra: dict) -> bytes:
        if len(idk_shares) == 0:
            raise ValueError("decrypt: no IDK shares provided")

        xs = [fr_mod(s.eval_point) for s in idk_shares]
        for i in range(len(xs)):
            for j in range(i + 1, len(xs)):
                if xs[i] == xs[j]:
                    raise ValueError("decrypt: duplicate evalPoint")

        lambdas: list[int] = []
        for i, xi in enumerate(xs):
            lam = 1
            for j, xj in enumerate(xs):
                if i == j:
                    continue
                lam = fr_mul(lam, fr_mul(fr_mod(-xj), fr_inv(fr_mod(xi - xj))))
            lambdas.append(lam)

        idk_full = None
        for i, share in enumerate(idk_shares):
            if lambdas[i] == 0:
                continue
            scaled = _ob.multiply(share.idk_share, lambdas[i])
            idk_full = scaled if idk_full is None else _ob.add(idk_full, scaled)
        if idk_full is None:
            raise ValueError("decrypt: all Lagrange coefficients were zero")

        seed = _seed_from_gt(ciphertext.c0, idk_full)
        mac_key = kdf(seed, DST_MAC, 32)
        mac_another = hmac_sha3_256(mac_key, ciphertext.symmetric_ciph)
        if ciphertext.mac != mac_another:
            raise ValueError("decrypt: MAC verification failed")
        otp = kdf(seed, DST_OTP, len(ciphertext.symmetric_ciph))
        return xor_bytes(otp, ciphertext.symmetric_ciph)

    return Result.capture(task)
